docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
+It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.
+
+## New APIs
+
+We add the new APIs for retreiving and interacting with nullability:
+
+```C#
+public enum Nullability
+{
+    NotApplicable,
+    NotComputed,
+    Unknown,
+    NotNull,
+    MayBeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IEventSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IFieldSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface ILocalSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IMethodSymbol : ISymbol
+{
+    Nullability ReturnNullability { get; }
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
+}
+
+public interface IParameterSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
+}
+
+public interface INamedTypeSymbol : ITypeSymbol
+{
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+}
+
+public interface ITypeParameterSymbol : ITypeSymbol
+{
+    Nullability ReferenceTypeConstraintNullability { get; }
+    ImmutableArray<Nullability> ConstraintsNullabilities { get; }
+}
+
+public struct TypeInfo
+{
+    public Nullability Nullability { get; }
+    public Nullability ConvertedNullability { get; }
+}
+
+#endregion
+
+#region Compilation
+
+public class Compilation
+{
+    public SemanticModel GetSemanticModel(bool skipNullabilityInformation);
+}
+
+
+#endregion
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
```

Comment:
This seems necessary to do now.  Especially so we can gauge how bad the viral nature of this new piece of data would be.

Reply:
As an exercise here, I went though `SyntaxGenerator` and came up with a list of functions that would need to be added, how many usages of the function there are currently, and how many would be avoided if we added `Nullability` to `ITypeSymbol`. I did not include any unit test usages in this list, or usages within the syntax generator and subclasses.

```
Declarations
FieldDeclaration - 1 overload, none avoided. 2 uses.
MethodDeclaration - 1 overload, none avoided. 1 uses.
OperatorDeclaration - 1 overload, none avoided. 2 uses.
ParameterDeclaration - 1 overload, none avoided. 2 uses.
PropertyDeclaration - 1 overload, none avoided. No uses outside unit tests.
IndexerDeclaration - 1 overload, none avoided. No uses outside unit tests.
EventDeclarartion - 1 overload, none avoided. No uses outside unit tests.
CustomEventDeclaration - 1 overload, none avoided. No uses outside unit tests.
DelegateDeclaration - 1 overload, none avoided. No uses outside unit tests.
Declaration - 1 overload, 1 avoided. No uses outside unit tests.
AliasImportDirective - 1 overload, 1 avoided. 1 use.

Would want to add -
GetNullability
WithNullability

Statements
LocalDeclarationStatement - 5 new overloads. Only 1 would be eliminted by adding nullability to ITypeReference. 7 references, two of which would not need updating if ITypeSymbol had nullability
CatchClause - 2 overloads, none avoided, no uses

Expressions:
Worth noting here is that most of the usages are of the API that would be avoided.
DefaultExpression - 2 overloads, 1 avoided. 3 usages
IdentifierName - 2 overloads, none avoided. At least 90 usages outside unit tests
Identifer - 1 overload, none avoided. 11 usages
NameExpression - 1 overload, 1 avoided. 1 usages
TypeExpression - 3 overloads, 2 avoided. 43 usages
ArrayTypeExpression - 1 overload, none avoided. No uses outside unit tests
TupleTypeExpression - 1 overload, 1 avoided. No uses outside unit tests
TupleElementExpression - 2 overloads, 1 avoided. 2 uses
ArrayCreationExpression - 2 overloads, none avoided. No uses outside unit tests
ObjectCreationExpression - 4 overloads, 2 avoided. 9 uses
IsTypeExpression - 1 overload, 1 avoided. 4 uses
TryCastExpressio - 1 overload, 1 avoided. 3 uses.
CastExpression - 1 overload, 1 avoided. 11 uses.
ConvertExpression - 1 overload, 1 avoided. 1 use.
LambdaParameter - 1 overload, 1 avoided. No uses outside tests.
```

Reply:
Could you clarify the data?  I'm not sure i undertaand it. For example, can you break down what these mean: 

```
FieldDeclaration - 1 overload, none avoided. 2 uses.
...
TypeExpression - 3 overloads, 2 avoided. 43 usages
```

(i.e. actually go into the specific for these two cases).

Reply:
Sure.
`Field Declaration - needs 1 overload, would not have been avoided by adding Nullability to ITypeSymbol. 2 uses of one of the existing APIs not in test code or in the generator`
`Type Expression - needs 3 overloads, 2 would have been avoided by adding Nullability to ITypeSymbol. 43 uses of one of the existing APIs not in test code or in the generator`

Reply:
Sorry, i was asking for more detail than that.  Specifically:

> Field Declaration - needs 1 overload, would not have been avoided by adding Nullability to ITypeSymbol.

What overload would you intent to add?

Same with the others.  Can you explain the overlaods you intend?  I ask because this is whati see:

```c#
SyntaxNode FieldDeclaration(
            string name,
            SyntaxNode type,
            Accessibility accessibility = Accessibility.NotApplicable,
            DeclarationModifiers modifiers = default,
            SyntaxNode initializer = null);
SyntaxNode FieldDeclaration(IFieldSymbol field);
SyntaxNode FieldDeclaration(IFieldSymbol field, SyntaxNode initializer);
```

I don't see why any of those would need overloads added.  They all seem sufficient for handling the new NRT feature.  Thanks!

Reply:
Yeah, you're right. Updated the declarations list to remove everything but `AliasImportDirective`, which will need the `AliasImportDeclaration(string aliasIdentifierName, INamespaceOrTypeSymbol symbol)` variant overloaded to take a `Nullability`, so you can say `using nullableType = Type?;`.

Reply:
> Yeah, you're right. Updated the declarations list to remove everything but AliasImportDirective

Why?  Since ITypeSymbol's would have the nullability info, that would come through `INamespaceOrTypeSymbol symbol`?

Reply:
> Updated the declarations list to remove everything

Why just the declarations list?  I'm not seeing why your other cases would need this either.  For example, from the 'expressions' i looked at, none seem like they would need overloads.  Can you clarify why you think they would?